### PR TITLE
coinbase check in snark

### PR DESCRIPTION
**Pull request by deepthiskumar**
Here's the version of the library to be used for tracking coinbases

### CODE REVIEW

File: src/lib/coda_base/pending_coinbase.ml [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+        let%bind () = Boolean.Assert.is_true yes in
+        return yes )
+      ~f:(fun is_empty_or_writeable x -> f ~is_empty_or_writeable x)
+
+  let%snarkydef delete_stack t ~f =
+    let filter stack =
+      let%bind empty_stack =
+        Stack.Checked.equal stack Stack.(var_of_t empty)
+      in
+      let%bind () = Boolean.(Assert.is_true (not empty_stack)) in
```
Comment (imeckler) [inline src/lib/coda_base/pending_coinbase.ml]: same as above

File: src/lib/coda_base/pending_coinbase_intf.ml [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    val merge : height:int -> t -> t -> t
+
+    val empty_hash : t
+
+    val of_digest : Pedersen.Digest.t -> t
+
+    val update_stack :
+         var
+      -> is_new_stack:Boolean.var
+      -> f:(   is_empty_or_writeable:Boolean.var
```
Comment (imeckler) [inline src/lib/coda_base/pending_coinbase_intf.ml]: what do these signify/why do we need them?

File: src/lib/coda_base/pending_coinbase.ml [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+*)
+  let update_stack t ~is_new_stack ~f =
+    modify_stack' t
+      ~filter:(fun stack ->
+        let%bind empty_stack =
+          Stack.Checked.equal stack Stack.(var_of_t empty)
+        in
+        let%bind new_stack = Boolean.(empty_stack && is_new_stack) in
+        let%bind yes = Boolean.(new_stack || not empty_stack) in
+        let%bind () = Boolean.Assert.is_true yes in
```
[the remark this answers is not in the document]
Comment (deepthiskumar) [reply src/lib/coda_base/pending_coinbase.ml]: Ah, no, it had another boolean there which I deleted.

File: src/lib/coda_base/pending_coinbase.ml (hunk 1 above)
Comment (deepthiskumar) [reply src/lib/coda_base/pending_coinbase.ml]: I added an assertion here to know early on if there would be an error. We should not be trying to delete an empty stack.

File: src/lib/coda_base/pending_coinbase_intf.ml (hunk 2 above)
Comment (deepthiskumar) [reply src/lib/coda_base/pending_coinbase_intf.ml]: We don't need them. I made this structure (similar to sparse ledger requests) before writing how this would be called. I've deleted it in the version I have now

File: src/lib/coda_base/pending_coinbase.ml [hunk 4]
```diff
@@ -0,0 +1,552 @@
+open Core_kernel
```
Comment (imeckler) [inline src/lib/coda_base/pending_coinbase.ml]: Reminder to myself to come back here (haven't read it yet)

File: src/lib/blockchain_snark/blockchain_state.ml [hunk 5]
```diff
@@ -63,15 +63,29 @@ module Make (Consensus_mechanism : Consensus.S) :
           Consensus_mechanism.next_state_checked ~prev_state:previous_state
             ~prev_state_hash:previous_state_hash transition supply_increase
         in
+        let pending_coinbase_update =
```
Comment (imeckler) [inline src/lib/blockchain_snark/blockchain_state.ml]: reminder to myself to read this carefully

File: src/app/cli/src/prover.ml [hunk 6]
```diff
@@ -267,9 +274,10 @@ let create ~conf_dir =
 let initialized {connection; _} =
   Worker.Connection.run connection ~f:Worker.functions.initialized ~arg:()
 
-let extend_blockchain {connection; _} chain next_state block prover_state =
+let extend_blockchain {connection; _} chain next_state block prover_state
+    pending_coinbase =
   Worker.Connection.run connection ~f:Worker.functions.extend_blockchain
-    ~arg:(chain, next_state, block, prover_state)
+    ~arg:(chain, next_state, block, prover_state, pending_coinbase)
```
Comment (imeckler) [inline src/app/cli/src/prover.ml]: maybe the pending coinbase should be in the prover state? What do you think?

File: src/lib/blockchain_snark/blockchain_state.ml [hunk 7]
```diff
@@ -82,10 +96,63 @@ module Make (Consensus_mechanism : Consensus.S) :
...
+                   ~else_:
+                     (chain Pending_coinbase.Hash.if_ del_add
+                        ~then_:(with_del_add del_add)
+                        ~else_:(with_del_update del_updated)))
+          in
+          let%bind correct_coinbase_status =
+            let new_root = pending_coinbase_update.new_root in
+            Pending_coinbase.Hash.equal_var new_pending_coinbase_hash new_root
+          in
           let%bind correct_snark =
```
Comment (imeckler) [inline src/lib/blockchain_snark/blockchain_state.ml]: I think this is might now be broken because there is other info we use from the SNARK besides the ledge hash (e.g., total supply etc.).

It would probably be better if we first compute `correct_snark` and then use it as a mask on all the fields that we update using data from the input of the ledger proof.

File: src/lib/blockchain_snark/blockchain_state.ml [hunk 8]
```diff
@@ -82,10 +96,63 @@ module Make (Consensus_mechanism : Consensus.S) :
               ( transition |> Snark_transition.blockchain_state
               |> Blockchain_state.snarked_ledger_hash )
           in
+          let%bind new_pending_coinbase_hash =
```
Comment (imeckler) [inline src/lib/blockchain_snark/blockchain_state.ml]: Reminder to come back here

File: src/lib/coda_base/pending_coinbase.ml [hunk 9]
```diff
@@ -0,0 +1,552 @@
...
+open Fold_lib
+open Snark_bits
+
+let coinbase_tree_depth = Snark_params.pending_coinbase_depth
+
+(*Total number of stacks*)
+let coinbase_stacks = Int.pow 2 coinbase_tree_depth
+
+module Coinbase_data = struct
+  type t = Public_key.Compressed.t * Amount.Signed.t [@@deriving sexp]
```
Comment (imeckler) [inline src/lib/coda_base/pending_coinbase.ml]: Reminder to myself to create an issue to reuse the hash section of the proposer public key.

Comment (imeckler) [inline src/lib/coda_base/pending_coinbase.ml]: Why is it a signed amount? Shouldn't it be unsigned?

File: src/lib/coda_base/pending_coinbase.ml [hunk 10]
```diff
@@ -0,0 +1,552 @@
...
+      var_to_triples t
+      >>= Pedersen.Checked.hash_triples ~init:crypto_hash_prefix
+
+    let digest t =
+      var_to_triples t
+      >>= Pedersen.Checked.digest_triples ~init:crypto_hash_prefix
+  end
+end
+
+module Stack_pos = struct
```
Comment (imeckler) [inline src/lib/coda_base/pending_coinbase.ml]: maybe put a type signature on this to make sure you don't accidentally turn a too-large int into a stack_pos?

File: src/lib/coda_base/pending_coinbase.ml [hunk 11]
```diff
@@ -0,0 +1,552 @@
...
+        var_of_hash_packed digest
+
+      let if_ = if_
+
+      let empty = var_of_t empty
+
+      let equal (x : var) (y : var) =
+        Field.Checked.equal (var_to_hash_packed x) (var_to_hash_packed y)
+
+      let hash t =
```
Comment (imeckler) [inline src/lib/coda_base/pending_coinbase.ml]: I would just make hash Fn.id

File: src/lib/coda_base/pending_coinbase.ml [hunk 12]
```diff
@@ -0,0 +1,552 @@
...
+        end)
+
+    let depth = coinbase_tree_depth
+
+    type path = Pedersen.Digest.t list
+
+    type index = int
+
+    type _ Request.t +=
+      | Coinbase_stack_path : index -> path Request.t
```
Comment (imeckler) [inline src/lib/coda_base/pending_coinbase.ml]: Maybe `Stack_pos.t` rather than `index`?

File: src/lib/coda_base/pending_coinbase.ml [hunk 13]
```diff
@@ -0,0 +1,552 @@
...
+      in
+      handle
+        (Merkle_tree.modify_req ~depth (Hash.var_to_hash_packed t) addr
+           ~f:(fun _s -> return stack ))
+        reraise_merkle_requests
+      >>| Hash.var_of_hash_packed
+
+    let%snarkydef update_stack t ~is_new_stack stack =
+      let%bind updated_tree_hash = update_stack' t ~is_new_stack stack in
+      let%map () = reset () in
```
Comment (imeckler) [inline src/lib/coda_base/pending_coinbase.ml]: reminder to check what reset does

File: src/lib/coda_base/pending_coinbase.ml [hunk 14]
```diff
@@ -0,0 +1,552 @@
...
+    let update_stack' t ~is_new_stack stack =
+      let%bind addr =
+        request_witness Stack_pos.Unpacked.typ
+          As_prover.(
+            map (read Boolean.typ is_new_stack) ~f:(fun s ->
+                Find_index_of_newest_stack s ))
+      in
+      handle
+        (Merkle_tree.modify_req ~depth (Hash.var_to_hash_packed t) addr
+           ~f:(fun _s -> return stack ))
```
Comment (imeckler) [inline src/lib/coda_base/pending_coinbase.ml]: isn't here where you're suppose to push the new coinbase?

File: src/lib/coda_base/pending_coinbase_intf.ml [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+    val hash : t -> Hash.t
+  end
+
+  val create_exn : unit -> t
+
+  val remove_coinbase_stack_exn : t -> t
+
+  val merkle_root : t -> Hash.t
+
+  val empty_merkle_root : unit -> Hash.t
```
Comment (imeckler) [inline src/lib/coda_base/pending_coinbase_intf.ml]: Why does this take unit?

File: src/lib/blockchain_snark/blockchain_state.ml [hunk 16]
```diff
@@ -82,10 +96,63 @@ module Make (Consensus_mechanism : Consensus.S) :
...
+            let prev_root =
+              previous_state |> Protocol_state.blockchain_state
+              |> Blockchain_state.pending_coinbase_hash
+            in
+            let updated_stack = pending_coinbase_update.updated_stack in
+            let action = pending_coinbase_update.action in
+            let with_del_add del_added =
```
Comment (imeckler) [inline src/lib/blockchain_snark/blockchain_state.ml]: do this in 2 modifies

File: src/lib/coda_base/pending_coinbase.ml (hunk 14 above)
Comment (deepthiskumar) [reply src/lib/coda_base/pending_coinbase.ml]: As discussed, coinbase is pushed to the stack in transaction snark. Here, the old stack is replaced with the new stack. I've now added a check for the old stack

File: src/lib/coda_base/pending_coinbase_intf.ml (hunk 15 above)
Comment (deepthiskumar) [reply src/lib/coda_base/pending_coinbase_intf.ml]: Because it first creates a merkle tree with all empty stacks.

File: src/lib/coda_base/pending_coinbase.ml [hunk 17]
```diff
@@ -0,0 +1,499 @@
...
+        let digest, _ =
+          Pedersen.Checked.Section.to_initial_segment_digest_exn s
+        in
+        var_of_hash_packed digest
+
+      let if_ = if_
+
+      let empty = var_of_t empty
+
+      let digest t =
```
Comment (imeckler) [inline src/lib/coda_base/pending_coinbase.ml]: Is this ever used? What for if so?

File: src/lib/coda_base/pending_coinbase.ml [hunk 18]
```diff
@@ -0,0 +1,499 @@
...
+          let assert_equal h1 h2 = Field.Checked.Assert.equal h1 h2
+
+          let if_ = Field.Checked.if_
+        end)
+        (struct
+          include Stack
+
+          type value = t [@@deriving sexp]
+
+          let hash = Checked.digest
```
Comment (imeckler) [inline src/lib/coda_base/pending_coinbase.ml]: This should be Fn.id (or whatever the necessary type coercion is.

File: src/lib/coda_base/pending_coinbase.ml [hunk 19]
```diff
@@ -0,0 +1,499 @@
...
+        (Merkle_tree.modify_req ~depth (Hash.var_to_hash_packed t) addr
+           ~f:(fun stack ->
+             let%map () =
+               Coinbase_stack.Stack.assert_equal stack_before stack
+             in
+             stack_after ))
+        reraise_merkle_requests
+      >>| Hash.var_of_hash_packed
+
+    let%snarkydef delete_stack t stack_before stack_after =
```
Comment (imeckler) [inline src/lib/coda_base/pending_coinbase.ml]: This function is essentially the same as the above so I would reuse the code. Also I don't think delete stack is necessarily a good name. What about this function actually involves deleting a stack?

File: src/lib/coda_base/pending_coinbase.ml [hunk 20]
```diff
@@ -0,0 +1,499 @@
...
+          Option.value_exn
+            (List.Assoc.find ~equal:Int.equal hash_on_level height)
+        in
+        create_path (height - 1)
+          ((if key mod 2 = 0 then `Left hash else `Right hash) :: path)
+          (key / 2)
+    in
+    let rec go t key =
+      if
+        Stack_id.compare key (Stack_id.of_int @@ Int.pow 2 coinbase_tree_depth)
```
Comment (imeckler) [inline src/lib/coda_base/pending_coinbase.ml]: I would use `Stack_id.(>)`, I think it's more clear than using compare directly

File: src/lib/coda_base/pending_coinbase.ml [hunk 21]
```diff
@@ -0,0 +1,499 @@
...
+             in
+             stack_after ))
+        reraise_merkle_requests
+      >>| Hash.var_of_hash_packed
+  end
+
+  type t = {tree: Merkle_tree.t; pos_list: Stack_id.t list; new_pos: Stack_id.t}
+  [@@deriving sexp, bin_io]
+
+  let create_exn () =
```
Comment (imeckler) [inline src/lib/coda_base/pending_coinbase.ml]: what does this function do?

File: src/lib/coda_base/pending_coinbase.ml [hunk 22]
```diff
@@ -0,0 +1,499 @@
...
+          Option.value_exn
+            ~error:(Error.of_string "Stack_id overflowed")
+            (Stack_id.incr_by_one t.new_pos)
+      in
+      {t with pos_list= t.new_pos :: t.pos_list; new_pos}
+    else t
+
+  let latest_stack_id t ~is_new_stack =
+    if is_new_stack then Some t.new_pos
+    else match t.pos_list with [] -> None | x :: _ -> Some x
```
Comment (imeckler) [inline src/lib/coda_base/pending_coinbase.ml]: You can use `List.hd`

File: src/lib/coda_base/pending_coinbase.ml [hunk 23]
```diff
@@ -0,0 +1,499 @@
...
+
+  let latest_stack_exn t =
```
Comment (imeckler) [inline src/lib/coda_base/pending_coinbase.ml]: I think it's confusing for an `_exn` function to return an option. I would use Option.try_with if you want to return an option here.

File: src/lib/coda_base/pending_coinbase_update.ml [hunk 24]
```diff
@@ -0,0 +1,154 @@
...
+  { oldest_stack_before
+  ; oldest_stack_after
+  ; latest_stack_before
+  ; latest_stack_after
+  ; is_new_stack
+  ; prev_root
+  ; intermediate_root
+  ; new_root }
+
+let typ : (var, t) Typ.t =
```
Comment (imeckler) [inline src/lib/coda_base/pending_coinbase_update.ml]: I am skeptical this is necessary

File: src/lib/coda_base/staged_ledger_hash.ml [hunk 25]
```diff
@@ -41,16 +41,144 @@ module Aux_hash = struct
   let to_bytes = Fn.id
 
   let dummy : t = String.init length_in_bytes ~f:(fun _ -> '\000')
+end
+
+module Pending_coinbase_extra = struct
```
Comment (imeckler) [inline src/lib/coda_base/staged_ledger_hash.ml]: Can we call this aux to at least be somewhat consistent with staged ledger?

File: src/lib/coda_base/staged_ledger_hash.ml [hunk 26]
```diff
@@ -41,16 +41,144 @@ module Aux_hash = struct
...
+  include Stable.Latest
+
+  let of_bytes = Fn.id
+
+  let to_bytes = Fn.id
+
+  let dummy : t = String.init length_in_bytes ~f:(fun _ -> '\000')
+end
+
+module Non_snark = struct
```
Comment (imeckler) [inline src/lib/coda_base/staged_ledger_hash.ml]: Let's see if we can come up with a better name

File: src/lib/coda_base/transaction_union.ml [hunk 27]
```diff
@@ -50,14 +50,11 @@ let of_transaction : Transaction.t -> t = function
...
-              { fee= Amount.to_fee amount
+              { fee= other_amount
               ; nonce= Account.Nonce.zero
               ; memo= User_command_memo.dummy }
-          ; body=
-              { public_key= other_pk
-              ; amount= Amount.of_fee other_amount
-              ; tag= Tag.Coinbase } }
-      ; sender= Public_key.decompress_exn proposer
+          ; body= {public_key= proposer; amount; tag= Tag.Coinbase} }
```
Comment (imeckler) [inline src/lib/coda_base/transaction_union.ml]: why was this changed?

File: src/lib/transaction_snark/transaction_snark.ml [hunk 28]
```diff
@@ -316,21 +364,32 @@ module Base = struct
       with_label __LOC__
         (exists' Transaction_union.typ ~f:Prover_state.transaction)
     in
-    let%bind root_after, fee_excess, supply_increase =
-      apply_tagged_transaction (module Shifted) root_before t
+    let%bind pending_coinbase_before =
```
Comment (imeckler) [inline src/lib/transaction_snark/transaction_snark.ml]: reminder to check this is correct

File: src/lib/blockchain_snark/blockchain_state.ml [hunk 29]
```diff
@@ -82,10 +96,35 @@ module Make (Consensus_mechanism : Consensus.S) :
               ( transition |> Snark_transition.blockchain_state
               |> Blockchain_state.snarked_ledger_hash )
           in
+          let%bind new_pending_coinbase_hash =
+            let%bind _root_after_delete =
```
Comment (imeckler) [inline src/lib/blockchain_snark/blockchain_state.ml]: Don't throw away -- pass into delete

File: src/lib/coda_base/pending_coinbase.ml (hunk 17 above)
Comment (deepthiskumar) [reply src/lib/coda_base/pending_coinbase.ml]: It is not, I removed it.

File: src/lib/coda_base/pending_coinbase.ml (hunk 21 above)
Comment (deepthiskumar) [reply src/lib/coda_base/pending_coinbase.ml]: Creates a merkle tree with empty stacks. This function should go probably in `Sparse_ledger`

File: src/lib/coda_base/pending_coinbase.ml (hunk 23 above)
Comment (deepthiskumar) [reply src/lib/coda_base/pending_coinbase.ml]: Changed all the `_exn` functions to return an `Or_error.t`

File: src/lib/coda_base/staged_ledger_hash.ml (hunk 26 above)
Comment (deepthiskumar) [reply src/lib/coda_base/staged_ledger_hash.ml]: `Unchecked_hash`?

File: src/lib/coda_base/transaction_union.ml (hunk 27 above)
Comment (deepthiskumar) [reply src/lib/coda_base/transaction_union.ml]: To have coinbase amount as an usigned int. I swapped fee and amount (receiver and sender accordingly). Moved this to another PR (#1902)

File: src/lib/blockchain_snark/blockchain_state.ml [hunk 30]
```diff
@@ -83,12 +88,11 @@ module Make (Consensus_mechanism : Consensus.S) :
             in
             (*new stack or update one*)
             let%bind coinbase_amount =
```
Comment (imeckler) [inline src/lib/blockchain_snark/blockchain_state.ml]: No need to do an exists here, you can just use `Currency.Amount.var_of_t`
